src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Components/ComponentBindLoweringPass.cs

```diff
@@ -213,120 +239,143 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentInte
-            // This is a workaround for a limitation where you can't write a tag helper that binds only
-            // when a specific attribute is **not** present.
-            if (attribute != null &&
-                tagHelper != null &&
-                tagHelper.IsInputElementFallbackBindTagHelper())
+        // Now that we've identified attributes that should be removed, iterate them in reverse order
+        // and remove them using the index we stored. Note: We know the attributes are already ordered
+        // by index.
+        for (var i = toRemove.Count - 1; i >= 0; i--)
+        {
+            children.RemoveAt(toRemove[i].Index);
+        }
+
+        // If we still have duplicates at this point then they are genuine conflicts.
+        // Use a hash set to quickly determine whether there are any duplicates.
+        // If so, we need to do a more expensive pass to identify and remove them.
+        using var _ = StringHashSetPool.Ordinal.GetPooledObject(out var duplicates);
+
+        foreach (var child in children)
+        {
+            if (child is TagHelperDirectiveAttributeIntermediateNode { AttributeName: string attributeName } &&
+                !duplicates.Add(attributeName))
             {
-                for (var j = 0; j < node.Children.Count; j++)
+                ReportDiagnosticAndRemoveDuplicates(node);
+                break;
+            }
+        }
+
+        static void ReportDiagnosticAndRemoveDuplicates(IntermediateNode node)
+        {
+            var children = node.Children;
+
+            using var _ = StringDictionaryPool<ImmutableArray<AttributeInfo>.Builder>.Ordinal.GetPooledObject(out var duplicates);
+
+            for (var i = 0; i < children.Count; i++)
+            {
+                if (children[i] is TagHelperDirectiveAttributeIntermediateNode directiveAttribute)
                 {
-                    TagHelperDescriptor duplicateTagHelper = null;
-                    string duplicateAttributeName = null;
-                    var duplicate = node.Children[j];
-                    if (duplicate is TagHelperDirectiveAttributeIntermediateNode duplicateDirectiveAttribute)
-                    {
-                        duplicateAttributeName = duplicateDirectiveAttribute.AttributeName;
-                        duplicateTagHelper = duplicateDirectiveAttribute.TagHelper;
-                    }
-                    else if (duplicate is TagHelperDirectiveAttributeParameterIntermediateNode duplicateParameterAttribute)
+                    if (!duplicates.TryGetValue(directiveAttribute.AttributeName, out var builder))
                     {
-                        duplicateAttributeName = duplicateParameterAttribute.AttributeName;
-                        duplicateTagHelper = duplicateParameterAttribute.TagHelper;
+                        builder = ImmutableArray.CreateBuilder<AttributeInfo>();
```

Comment:
>uilder = ImmutableArray.CreateBuilder<AttributeInfo>(); [](http://example.com/codeflow?start=25&length=55)

Feels like we're doing this more than potentially necessary. If the caller to this method collected all duplicates and then passed in it's duplicate hashset, couldn't we only create the IA.Builders for the actual duplicated attributes?

Reply:
I didn't feel it was worth optimizing the error scenario, since that would be less common.